### PR TITLE
fix: Frame.Navigating cancellation, feat: Frame.IsNavigationStackEnabled

### LINKED ISSUE #2893: NavigatingFrom called on Page even if Frame.Navigating cancels
## Current behavior

When `Frame.Navigating` sets `Cancel`, the current page's `NavigatingFrom` is still called.

## Expected behavior

When `Frame.Navigating` sets `Cancel`, the current page's `NavigatingFrom` should not be called and `Frame.NavigationStopped` should follow immediately.

## How to reproduce it (as minimally and precisely as possible)

Wire up `Page.NavigatingFrom` and `Frame.Navigating`. In `Navigating` handler set `Cancel` to `true`.

## Environment

Nuget Package: latest

Package Version(s): latest

Affected platform(s):

- [x] iOS
- [x] Android
- [x] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [x] macOS
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

Visual Studio:

- [ ] 2017 (version: )
- [ ] 2019 (version: )
- [ ] for Mac (version: )

Relevant plugins:

- [ ] Resharper (version: )

### LINKED ISSUE #2928: Support Frame.IsNavigationStackEnabled
## What would you like to be added:

`Frame.IsNavigationStackEnabled` property.

## Why is this needed:

Stopping navigation tracking.

## For which Platform:

- [x] iOS
- [x] Android
- [x] WebAssembly
- [x] WebAssembly renderers for Xamarin.Forms
- [x] macOS
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

### PR DESCRIPTION
**Pull request by MartinZikmund**
GitHub Issue (If applicable): closes #2928, fixes #2893

## PR Type

What kind of change does this PR introduce?

- Bugfix
- Feature

## What is the current behavior?

- `Frame.Navigating` cancellation doesn't prevent `OnNavigatingFrom` from executing.
- `Frame.IsNavigationStackEnabled` is not available.

## What is the new behavior?

- `Frame.Navigating` can cancel navigation.
- `Frame.IsNavigationStackEnabled` is available.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [x] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

### CODE REVIEW

File: src/Uno.UI.Tests/Windows_UI_XAML_Controls/Frame/Given_Frame.cs [hunk 1]
```diff
@@ -57,4 +173,26 @@ public void When_RemovedPage()
...
+	{
+	}
+
+	class ThirdPage : Page
+	{
+	}
+
+	class DisallowNavigatingFromPage : Page
+	{
+		public static bool NavigatingFromCalled = false;
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI.Tests/Windows_UI_XAML_Controls/Frame/Given_Frame.cs:191]: This is intentional.

**Comment by jeromelaban** [comment]
You may want to reword `feat: Support for IsNavigationStackEnabled` to `feat: Support for Frame.IsNavigationStackEnabled`

**Comment by MartinZikmund** [comment]
@jeromelaban will do, are the release notes already generated somewhere?

**Comment by jeromelaban** [comment]
> @jeromelaban will do, are the release notes already generated somewhere?

Not yet, we're still looking for the right tool to generate those.

File: src/Uno.UI.Tests/Windows_UI_XAML_Controls/Frame/Given_Frame.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UI.Tests/Windows_UI_XAML_Controls/Frame/Given_Frame.cs:191]: Intentional

**Comment by MartinZikmund** [comment]
@jeromelaban I'm getting ` Error : Removed property Windows.UI.Xaml.DependencyProperty Windows.UI.Xaml.Controls.Frame::IsNavigationStackEnabledProperty() not found in ignore set.`, is that because the generated class was missing the `readonly` modifier or am I missing something?

**Comment by davidjohnoliver** [comment]
> @jeromelaban I'm getting ` Error : Removed property Windows.UI.Xaml.DependencyProperty Windows.UI.Xaml.Controls.Frame::IsNavigationStackEnabledProperty() not found in ignore set.`, is that because the generated class was missing the `readonly` modifier or am I missing something?

@MartinZikmund It's because the one you added is actually a field, add a `{get;} =` ;)

**Comment by MartinZikmund** [comment]
> > @jeromelaban I'm getting ` Error : Removed property Windows.UI.Xaml.DependencyProperty Windows.UI.Xaml.Controls.Frame::IsNavigationStackEnabledProperty() not found in ignore set.`, is that because the generated class was missing the `readonly` modifier or am I missing something?
> 
> @MartinZikmund It's because the one you added is actually a field, add a `{get;} =` ;)
> 
> (replacing a property with a field is a binary breaking change)

But the `DependencyProperty` should be fields I think (even all the others in `Frame`  and in other Uno.UI classes are just fields), is it possible the generated class is wrong?

**Comment by MartinZikmund** [comment]
@davidjohnoliver @jeromelaban That's weird. Seems like WPF's `DependencyProperty` are all fields, but in UWP they are read-only properties. Unfortunately VS generates it as a field, so Uno has most of them as fields instead of properties 😮 . See [WPF Button](https://docs.microsoft.com/cs-cz/dotnet/api/system.windows.controls.button?view=netframework-4.8) vs [UWP Button](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Controls.Button)

File: src/Uno.UI/UI/Xaml/Controls/Frame/Frame.cs [hunk 2]
```diff
@@ -192,6 +192,19 @@ private void OnSourcePageTypeChanged(DependencyPropertyChangedEventArgs e)
...
 
+		#region IsNavigationStackEnabled DependencyProperty
+
+		public bool IsNavigationStackEnabled
+		{
+			get { return (bool)GetValue(IsNavigationStackEnabledProperty); }
+			set { SetValue(IsNavigationStackEnabledProperty, value); }
+		}
+
+		public static readonly DependencyProperty IsNavigationStackEnabledProperty =
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/Frame/Frame.cs:203]:
```suggestion
		public static readonly DependencyProperty IsNavigationStackEnabledProperty =>
```

Don't change the property to a field

Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Controls/Frame/Frame.cs:203]: I know but all other DependencyProperty declarations are fields 🤔

Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Controls/Frame/Frame.cs:203]: I could make this one a property and the we could add a task to convert every "field" DependencyProperty in the codebase to property to the breaking change rollup.

Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Controls/Frame/Frame.cs]: Made #2953 for this

**Pull request merged**